### PR TITLE
[next] Fix mesh TextureMatrix and CanvasMeshRenderer

### PR DESCRIPTION
**Pull request by ivanpopelyshev**
Test it here: https://www.pixiplayground.com/#/edit/b0RAyPYeQobRMwa2aMQ~T

Build: https://pixijs.download/next-mesh-fix/pixi-legacy.js

Unlike default rope example, this thing also changes the texture matrix every frame.

Things to check:

1. `forceCanvas`
2. `batchable`
3. `app.stage.alpha`
4. `tick += 0.1`
5. `tick2 += 0.1;`
6. `PIXI.settings.MESH_CANVAS_PADDING`

### CODE REVIEW

**Comment by bigtimebuddy** [comment]
Awesome work @ivanpopelyshev

File: packages/mesh/src/MeshMaterial.js [hunk 1]
```diff
@@ -17,19 +17,32 @@ export default class MeshMaterial extends Shader
...
 
-        super(program, uniforms);
+        // Set defaults
+        options = Object.assign({
+            tint: 0xFFFFFF,
+            alpha: 1,
+            pluginName: 'batch',
+        }, options);
+
+        if (options.uniforms)
```
Comment (bigtimebuddy) [inline packages/mesh/src/MeshMaterial.js]: You don't need the `if` check here, doing `Object.assign({}, undefined)` is fine

File: packages/mesh/src/MeshBatchUvs.js [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+ */
+export default class MeshBatchUvs
+{
+    /**
+     * @param {PIXI.Buffer} uvBuffer - Buffer with normalized uv's
+     * @param {PIXI.TextureMatrix} uvMatrix - Material UV matrix
+     */
+    constructor(uvBuffer, uvMatrix)
+    {
+        this.uvBuffer = uvBuffer;
```
Comment (bigtimebuddy) [inline packages/mesh/src/MeshBatchUvs.js]:
```suggestion
        /**
         * Buffer with normalized UV's
         * @member {PIXI.Buffer}
         */
        this.uvBuffer = uvBuffer;
```

File: packages/mesh/src/MeshBatchUvs.js [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+        this.uvMatrix = uvMatrix;
```
Comment (bigtimebuddy) [inline packages/mesh/src/MeshBatchUvs.js]:
```suggestion
        /**
         * Material UV matrix
         * @member {PIXI.TextureMatrix}
         */
        this.uvMatrix = uvMatrix;
```

File: packages/mesh/src/MeshBatchUvs.js [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+        this.data = null;
```
Comment (bigtimebuddy) [inline packages/mesh/src/MeshBatchUvs.js]:
```suggestion
        /**
         * UV Buffer data
         * @member {Float32Array}
         * @readonly
         */
        this.data = null;
```

File: packages/core/src/textures/TextureMatrix.js [hunk 5]
```diff
@@ -62,6 +62,12 @@ export default class TextureMatrix
          * @member {number}
          */
         this.clampMargin = (typeof clampMargin === 'undefined') ? 0.5 : clampMargin;
+
+        /**
+         * If texture size is the same as baseTexture
+         * @readonly
```
Comment (bigtimebuddy) [inline packages/core/src/textures/TextureMatrix.js]:
```suggestion
         * @member {boolean}
         * @default false
         * @readonly
```

File: packages/canvas/canvas-mesh/src/CanvasMeshRenderer.js [hunk 6]
```diff
@@ -122,8 +122,8 @@ export default class CanvasMeshRenderer
     {
         const context = this.renderer.context;
         const vertices = mesh.geometry.buffers[0].data;
-        const uvs = mesh.geometry.buffers[1].data;
-        const texture = mesh._texture;
+        const uvs = mesh.uvs;
+        const texture = mesh.texture;
```
Comment (bigtimebuddy) [inline packages/canvas/canvas-mesh/src/CanvasMeshRenderer.js]:
```suggestion
        const {uvs, texture} = mesh;
```

File: packages/mesh/src/MeshMaterial.js (hunk 1 above)
Comment (ivanpopelyshev) [reply packages/mesh/src/MeshMaterial.js]: We are checking uniforms here, and I do not want to create those extra objects for them to be overwritten.

File: packages/canvas/canvas-mesh/src/CanvasMeshRenderer.js [hunk 7]
```diff
@@ -122,8 +122,8 @@ export default class CanvasMeshRenderer
     {
         const context = this.renderer.context;
         const vertices = mesh.geometry.buffers[0].data;
-        const uvs = mesh.geometry.buffers[1].data;
-        const texture = mesh._texture;
+        const uvs = mesh.uvs;
```
Comment (bigtimebuddy) [inline packages/canvas/canvas-mesh/src/CanvasMeshRenderer.js]:
```suggestion
```

Comment (bigtimebuddy) [reply packages/canvas/canvas-mesh/src/CanvasMeshRenderer.js]: Don’t need this anymore

**Pull request merged**